### PR TITLE
Added multitenancy for storage backend

### PR DESCRIPTION
**Pull request by MarioDegen**
Description of changes:
Forward tenantId to the dynamodb table.
The table key is an composite key (resourceid + tenantId) if multitenancy is enabled, tenantid is also separately stored in a column.
Utility functions are implemented to create the composite key and to extract the resource id and tenant id again.
For ddb to elasticsearch the resource id and tenant id are stored separately. The tenant id is also configured as index.
Added tests
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/objectStorageService/s3DataService.test.ts [hunk 1]
```diff
@@ -24,6 +24,106 @@ import { S3DataService } from './s3DataService';
 
 jest.mock('./s3ObjectStorageService');
 
+async function runCreateV4Test(s3DataService: S3DataService, tenantId?: string) {
```
Comment (ilyash-b) [inline src/objectStorageService/s3DataService.test.ts:27]: With and without tenantId tests are factored out - good!

**Review by ilyash-b — CHANGES_REQUESTED** [review]
* See comments
* Can (resource id + resource id) aspect be done more clear somehow?

File: src/dataServices/dynamoDbUtil.ts [hunk 2]
```diff
@@ -11,27 +11,60 @@ import DOCUMENT_STATUS from './documentStatus';
...
         delete cleanedItem[DOCUMENT_STATUS_FIELD];
         delete cleanedItem[LOCK_END_TS_FIELD];
         delete cleanedItem[VID_FIELD];
         delete cleanedItem[REFERENCES_FIELD];
-
+        if (DynamoDbUtil.hasTenantId(cleanedItem)) {
+            DynamoDbUtil.cleanItemId(cleanedItem);
+            if (!(projectionExpression !== undefined && projectionExpression.search(TENANT_ID_FIELD) !== -1)) {
+                delete cleanedItem[TENANT_ID_FIELD];
+            }
```
[the remark this answers is not in the document]
Comment (MarioDegen) [reply src/dataServices/dynamoDbUtil.ts:39]: During transaction handling it is necessary that getMostRecentResources returns the resource including the tenantid, with can be controlled by the projection expression

**Review by ilyash-b — CHANGES_REQUESTED** [review]
The amount of changes is huge. I strongly recommend different tradeoff: a bit more magic, much less code.

* Change resource `id` from `string` (primitive type) to `String` object. This allows arbitrary fields on the instances.
* Transform `id` and `tenantId` to new (concatenated or whatever) `id` as soon as possible, this will make the change transparent to most of the code
* Pass tenant id and original `id` along with the new (concatenated) `id`
* The above will eliminate most of the occurrences of `tenantId` and the need to pass it
* The places that need the original id and tenant id (few, low level places in the code)
* There can be several places in the code where this will not go smooth, talk to me if needed

See the code below for more information:
```
// Allow .fhirWorksTenantId on the built-in String type
interface String {
    fhirWorksOriginalResourceId: string;
    fhirWorksTenantId?: string;
}

// demo function with "string" type parameter
function f(s:string) {
}

f('aaa'); // works ok, argument type matches parameter type "string"

// vscode did not show any warning about the code line below
// https://www.jetbrains.com/help/idea/disabling-and-enabling-inspections.html#suppress-inspections
// noinspection JSPrimitiveTypeWrapperUsage
const s = new String('aaa');

s.fhirWorksTenantId = 'blah'; // works because of the interface definition at the top of the file

// f(s); // tsc: error TS2345: Argument of type 'String' is not assignable to parameter of type 'string'

f(s as string); // type cast, also works

// Solution rough idea - start
class DynamoDbUtil {
    // should probably move, the method is more generic than DynamoDb
    static buildItemId(resourceId: string, tenantId?: string): string {
        // noinspection JSPrimitiveTypeWrapperUsage
        const ret = new String(resourceId + (tenantId || ''));
        ret.fhirWorksOriginalResourceId = resourceId;
        ret.fhirWorksTenantId = tenantId;
        return ret as string;
    }
}

const id = DynamoDbUtil.buildItemId('rrr', 'ttt');
console.log(JSON.stringify(id)); // correct, "rrrttt"
console.log(id.fhirWorksOriginalResourceId);
console.log(id.fhirWorksTenantId);

// Solution rough idea - end
```

**Comment by MarioDegen** [comment]
Hi @ilyash-b this is approach makes more sense. Is there also typescript an option the override cast operators, like in c++? So that f(s); would also work. Thx a lot for that feedback!
Next week I try it out and start with snkeos/fhir-works-on-aws-interface.

**Comment by ilyash-b** [comment]
> Is there also typescript an option the override cast operators, like in c++? So that f(s); would also work.

Not aware of such thing

**Pull request closed without merging**